### PR TITLE
Use 'odo delete` to delete a component that's running locally with odo

### PR DESCRIPTION
**Pull request by jgwest**
**What type of PR is this?**
/kind feature

**What does does this PR do / why we need it**:
This PR adds support to the `odo delete` command to allow it to delete odo components running in local docker containers.

**Which issue(s) this PR fixes**:
https://github.com/openshift/odo/issues/2581

**How to test changes / Special notes to the reviewer**:

See https://github.com/openshift/odo/pull/2806 for how to enable a local docker push target and create a component.

### CODE REVIEW

File: glide.yaml [hunk 1]
```diff
@@ -78,5 +78,8 @@ import:
   version: v19.03.8
 - package: github.com/containerd/containerd
   version: v1.3.3
+- package: github.com/golang/mock
```
Comment (johnmcollier) [inline glide.yaml:81]: 👍

Comment (johnmcollier) [inline glide.yaml:81]: 👍

File: pkg/devfile/adapters/docker/component/adapter.go [hunk 2]
```diff
@@ -45,7 +48,64 @@ func (a Adapter) DoesComponentExist(cmpName string) bool {
...
+		if len(vols) == 0 {
+			return fmt.Errorf("unable to find source volume for component %s", componentName)
+		}
+
+		for _, vol := range vols {
+
+			// If the volume was found to be attached to the component's container, then delete the volume.
+			if _, ok := volumeNames[vol.Name]; ok {
+				err := a.Client.RemoveVolume(vol.Name)
+				if err != nil {
```
Comment (johnmcollier) [inline pkg/devfile/adapters/docker/component/adapter.go:100]: Will this be an issue if another container is mounting the volume also, but it hasn't been stopped yet?

Would it be better to:
1) Have a list of unique volumes
2) Iterate over each container:
     a) Retrieve the container's volumes and update the list of unique volumes
     b) Remove the container
3) Remove the unique volumes found in step 2

File: glide.yaml [hunk 3]
```diff
@@ -78,5 +78,8 @@ import:
...
+  version: v1.4.3
+
   subpackages:
```
[the remark this answers is not in the document]
Comment (jgwest) [reply glide.yaml:81]: 🤔... fixed😆

File: pkg/devfile/adapters/docker/component/adapter.go (hunk 2 above)
Comment (jgwest) [reply pkg/devfile/adapters/docker/component/adapter.go]: Agreed and updated 👍 .

File: pkg/devfile/adapters/docker/component/adapter.go [hunk 4]
```diff
@@ -45,7 +49,72 @@ func (a Adapter) DoesComponentExist(cmpName string) bool {
...
+	volumeLabels := utils.GetProjectVolumeLabels(componentName)
+	vols, err := a.Client.GetVolumesByLabel(volumeLabels)
+	if err != nil {
+		return errors.Wrapf(err, "unable to retrieve source volume for component "+componentName)
+	}
+	if len(vols) == 0 {
+		return fmt.Errorf("unable to find source volume for component %s", componentName)
+	}
+
+	volumesToDelete := map[string] /* volume name -> volume name (value unused)*/ string{}
```
Comment (cdrage) [inline pkg/devfile/adapters/docker/component/adapter.go:80]: forgot to remove the comment?

File: pkg/devfile/adapters/docker/component/adapter.go [hunk 5]
```diff
@@ -45,7 +49,72 @@ func (a Adapter) DoesComponentExist(cmpName string) bool {
...
+
+	for _, container := range componentContainer {
+
+		glog.V(3).Infof("Deleting container %s for component %s", container.ID, componentName)
```
Comment (cdrage) [inline pkg/devfile/adapters/docker/component/adapter.go:84]: I think you mean `V(4)` ? We don't use V(3)

File: pkg/devfile/adapters/docker/component/adapter.go [hunk 6]
```diff
@@ -45,7 +49,72 @@ func (a Adapter) DoesComponentExist(cmpName string) bool {
...
+			// to the deletion list.
+			if _, ok := volumeNames[vol.Name]; ok {
+				volumesToDelete[vol.Name] = vol.Name
+			}
+		}
+	}
+
+	// Finally, delete the volumes we discovered during container deletion.
+	for name := range volumesToDelete {
+		glog.V(3).Infof("Deleting the volume %s for component %s", name, componentName)
```
Comment (cdrage) [inline pkg/devfile/adapters/docker/component/adapter.go:111]: `V(4)`

File: pkg/devfile/adapters/docker/component/adapter_test.go [hunk 7]
```diff
@@ -129,3 +133,123 @@ func TestDoesComponentExist(t *testing.T) {
...
+	tests := []struct {
+		name              string
+		args              args
+		componentName     string
+		componentExists   bool
+		skipContainerList bool
+		wantErr           bool
+	}{
+		{
+			name: "case 1: component exists and given labels are valid",
```
Comment (cdrage) [inline pkg/devfile/adapters/docker/component/adapter_test.go:150]: nitpick, but capitalize `Case`

File: pkg/lclient/containers.go [hunk 8]
```diff
@@ -74,6 +74,15 @@ func (dc *Client) RemoveContainer(containerID string) error {
 	return nil
 }
 
+// RemoveVolume removes a volume with the specified volume ID
+func (dc *Client) RemoveVolume(volumeID string) error {
```
Comment (cdrage) [inline pkg/lclient/containers.go:78]: I see no unit tests written for this function :(

File: pkg/devfile/adapters/docker/component/adapter.go (hunk 4 above)
Comment (jgwest) [reply pkg/devfile/adapters/docker/component/adapter.go]: It was intentional (I often use this format to document map key/value, eg `/* key description -> value description*/`), but I removed it and replaced it with a more standard description of the variable.

File: pkg/devfile/adapters/docker/component/adapter.go (hunk 5 above)
Comment (jgwest) [reply pkg/devfile/adapters/docker/component/adapter.go]: Done

File: pkg/devfile/adapters/docker/component/adapter_test.go (hunk 7 above)
Comment (jgwest) [reply pkg/devfile/adapters/docker/component/adapter_test.go]: Nitpicks are always good 😄 , done.

File: pkg/lclient/containers.go (hunk 8 above)
Comment (jgwest) [reply pkg/lclient/containers.go:80]: Good catch, unit tests have been added to `containers_test.go`.

**Comment by cdrage** [comment]
/retest

Tests unfortunately are still failing for this PR. Let's see if it's a flake.

**Comment by kadel** [comment]
What about volumes created based on the `volumes` definition in devfile.yaml component?
Currently, they are not being deleted

```
▶ odo push
 •  Push devfile component backend  ...
 ✓  Pulling image quay.io/eclipse/che-java11-maven:nightly [5s]
 ✓  Starting container for quay.io/eclipse/che-java11-maven:nightly [382ms]
 ✓  Pulling image busybox [3s]
 ✓  Starting container for busybox [496ms]
 ✓  Push devfile component backend [9s]
 ✓  Changes successfully pushed to component

tmp/odo-examples/demo-backend  master ✗                                                                                                                                          14d ⚑ ◒
▶ docker volume ls
DRIVER              VOLUME NAME
local               m2-backend-nwpm
local               odo-project-source-backend-bqfz

tmp/odo-examples/demo-backend  master ✗                                                                                                                                          14d ⚑ ◒
▶ odo delete
? Are you sure you want to delete the devfile component: backend? Yes
 ✓  Deleting devfile component backend [398ms]
 ✓  Successfully deleted component

tmp/odo-examples/demo-backend  master ✗                                                                                                                                          14d ⚑ ◒
▶ docker volume ls
DRIVER              VOLUME NAME
local               m2-backend-nwpm
```

**Comment by jgwest** [comment]
@kadel The PR that added support for non-source voumes had not been merged at the time of completion for this PR. A separate PR under the feature issue will be created to handle that (fairly trivial) work.

**Comment by kadel** [comment]
> @kadel The PR that added support for non-source voumes had not been merged at the time of completion of this PR. A separate PR under the feature issue will be created to handle that (fairly trivial) work of deleting non-source volumes.

It is starting to get a little bit messy :-( if `odo create` creates something `odo delete` should delete it.

**Pull request merged**